File: samples/python/ws_trades.py
import os
import json
import signal
import sys
from datetime import datetime, timezone
from websocket import WebSocketApp
from dotenv import load_dotenv

load_dotenv()

GEMINI_WS_URL = os.getenv('GEMINI_WS_URL', 'wss://ws.gemini.com')


def stream_trades(symbol: str):
    """Stream real-time trades for a given symbol."""

    def on_open(ws):
        print(f'Connected to {GEMINI_WS_URL}')
        print(f'Subscribing to {symbol} trades...\n')

        ws.send(json.dumps({
            'id': '1',
            'method': 'subscribe',
            'params': [f'{symbol}@trade'],
        }))

    def on_message(ws, data):
        msg = json.loads(data)

        # Skip subscription confirmations
        if 'result' in msg or 'id' in msg:
            return

        side = 'sell' if msg['m'] else 'buy '
        time = datetime.fromtimestamp(msg['E'] / 1_000_000_000, tz=timezone.utc).isoformat()
        print(f"[{time}] {msg['s']} {side} {msg['q']} @ {msg['p']}")

    def on_error(ws, err):
        print(f'WebSocket error: {err}')

    def on_close(ws, status_code, msg):
        print('Connection closed')

    ws = WebSocketApp(
        GEMINI_WS_URL,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close,
    )

    # Graceful shutdown
    signal.signal(signal.SIGINT, lambda *_: (print('\nClosing connection...'), ws.close(), sys.exit(0)))

    ws.run_forever()


if __name__ == '__main__':
    # Show help message
    if '--help' in sys.argv or '-h' in sys.argv:
        print('Usage: python3 ws_trades.py [symbol]')
        print('Example: python3 ws_trades.py ethusd')
        print('Default symbol: btcusd')
        sys.exit(0)

    # Get symbol from command line or use default
    symbol = sys.argv[1] if len(sys.argv) > 1 else 'btcusd'
    stream_trades(symbol)

File: samples/typescript/src/wsTrades.ts
import WebSocket from 'ws';
import * as dotenv from 'dotenv';

dotenv.config();

const GEMINI_WS_URL = process.env.GEMINI_WS_URL || 'wss://ws.gemini.com';

function streamTrades(symbol: string): void {
  const ws = new WebSocket(GEMINI_WS_URL);

  ws.on('open', () => {
    console.log(`Connected to ${GEMINI_WS_URL}`);
    console.log(`Subscribing to ${symbol} trades...\n`);

    ws.send(JSON.stringify({
      id: '1',
      method: 'subscribe',
      params: [`${symbol}@trade`],
    }));
  });

  ws.on('message', (data: WebSocket.Data) => {
    const msg = JSON.parse(data.toString());

    // Skip subscription confirmations
    if (msg.result || msg.id) return;

    const side = msg.m ? 'sell' : 'buy ';
    const time = new Date(msg.E / 1_000_000).toISOString();
    console.log(`[${time}] ${msg.s} ${side} ${msg.q} @ ${msg.p}`);
  });

  ws.on('error', (err: Error) => {
    console.error('WebSocket error:', err.message);
  });

  ws.on('close', () => {
    console.log('Connection closed');
  });

  // Graceful shutdown
  process.on('SIGINT', () => {
    console.log('\nClosing connection...');
    ws.close();
    process.exit(0);
  });
}

// Get symbol from command line or use default
const symbol = process.argv[2] || 'btcusd';

if (process.argv.includes('--help') || process.argv.includes('-h')) {
  console.log('Usage: npx ts-node src/wsTrades.ts [symbol]');
  console.log('Example: npx ts-node src/wsTrades.ts ethusd');
  console.log('Default symbol: btcusd');
  process.exit(0);
}

streamTrades(symbol);
